cs/test/FasterLogTests.cs

```diff
@@ -13,6 +13,33 @@
 
 namespace FASTER.test
 {
+    [TestFixture]
+    internal class FasterLogStandAloneTests
+    {
+
+        [Test]
+        public void TestDisposeReleasesFileLocksWithInprogressCommit()
+        {
+            string commitPath = TestContext.CurrentContext.TestDirectory + "/" + TestContext.CurrentContext.Test.Name + "/";
+            DirectoryInfo di = Directory.CreateDirectory(commitPath);
+            IDevice device = Devices.CreateLogDevice(commitPath + "testDisposeReleasesFileLocksWithInprogressCommit.log", preallocateFile: true, deleteOnClose: false);
+            FasterLog fasterLog = new FasterLog(new FasterLogSettings { LogDevice = device, LogChecksum = LogChecksumType.PerEntry });
+            Assert.IsTrue(fasterLog.TryEnqueue(new byte[100], out long beginAddress));
+            fasterLog.Commit(spinWait: false);
+            fasterLog.Dispose();
+            device.Dispose();
+            while (true)
+            {
+                try
+                {
+                    di.Delete(recursive: true);
+                    break;
+                }
+                catch { }
```

Comment:
interesting, we still expect an exception? 
Did I get it wrong that after the current fix there'll be no exceptions in DirectoryInfo.Delete() ?

Reply:
This sequence, though rare, could happen:
1. Async IO thread calls `Commit`, which first checks `_disposed` to be false
2. `Dispose` sets `_disposed` to be true and disposes the file handle
3. `Commit` sees the handle is null, creates a new handle
4. Main thread tries to delete directory, fails
5. `Commit` thread re-checks `_dispose` and deletes the handle it just created (for handle reclamation correctness).

To avoid this case, we retry the directory deletion (it did not happen in the tests though). Note that I removed the `WaitForCommit` and the check for `IsFaulted` state in this testcase -- if you had that then this loop would not be needed.

Also, unrelated, we did seem to notice earlier (though not confirmed) that the OS releases the handle a bit later after the dispose is called, i.e. an immediate directory deletion does not always work. This is unclear though.